newrelic/api/llm_custom_attributes.py

```diff
@@ -0,0 +1,56 @@
+# Copyright 2010 New Relic, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import logging
+
+from newrelic.api.time_trace import TimeTrace, current_trace
+from newrelic.api.transaction import current_transaction
+from newrelic.common.async_wrapper import async_wrapper as get_async_wrapper
+from newrelic.common.object_names import callable_name
+from newrelic.common.object_wrapper import FunctionWrapper, wrap_object
+from newrelic.core.function_node import FunctionNode
+
+_logger = logging.getLogger(__name__)
+
+
+class WithLlmCustomAttributes(object):
+    def __init__(self, custom_attr_dict):
+        transaction = current_transaction()
+        if not isinstance(custom_attr_dict, dict) or custom_attr_dict is None:
+            raise TypeError("custom_attr_dict must be a dictionary. Received type: %s" % type(custom_attr_dict))
+
+        # Add "llm." prefix to all keys in attribute dictionary
+        prefixed_attr_dict = {}
+        for k, v in custom_attr_dict.items():
+            if not k.startswith("llm."):
+                _logger.warning("Invalid attribute name %s. Renamed to llm.%s." % (k, k))
```

Comment:
I wonder if we should be logging a warning at all here. It might get a bit cumbersome logging-wise and I think it's totally fine if the customer wants to let us convert the names to be `llm.` prefixed for them. We can just document it as part of the api.

Reply:
I tend to agree with this. I originally added this logging statement to match the Node implementation but I would be fine to take it out.